### DISCUSSION TITLE
Halide OpenCL questions

### DISCUSSION THREAD

**Discussion by harishch4** [halide/Halide#5445] [Q&A] [answered]
I understood that Halide has a OpenCL backend. Can someone clarify me on the below questions. 

1. Does it mean the Halide compiler converts the Halide code to OpenCL host and kernel? 
2. Is there a way to dump the host and kernel code to a file? 
3. How can i choose the platform and device in host code?

**Comment by harishch4** [comment]
I'm able to print the kernel code by using HL_DEBUG_CODEGEN=1. Looking for other answers.

**Comment by abadams** [comment, marked as answer]
Yes, the Halide compiler converts the halide code to a mix of host and kernel code, depending on the schedule. The kernel code is OpenCL C, and there's no good way to dump it except for the one you already discovered. The host code compiles from Halide, to LLVM IR, to machine code. It never exists as C source code. You can see it in Halide IR and LLVM IR form with HL_DEBUG_CODEGEN=2.

To choose the platform and device see the functions in HalideRuntimeOpenCL.h (halide_opencl_set_platform_name and halide_opencl_set_device_type)

**Reply by abadams** [reply]
Oh, note those functions are only usable if you're using Halide in AOT-compiled mode. The environment variables mentioned in the doc strings for those functions still work in JIT mode though.

**Reply by harishch4** [reply]
Thanks @abadams, It helped.
